code/game/objects/items/motion_detector.dm

```diff
@@ -1,183 +1,64 @@
-	user?.client?.images -= blip_image
-	qdel(src)
-
-/obj/effect/blip/close_blip/Destroy()
-	blip_image = null
-	return ..()
-
 /obj/item/attachable/motiondetector
 	name = "tactical sensor"
 	desc = "A device that detects hostile movement. Hostiles appear as red blips. Friendlies with the correct IFF signature appear as green, and their bodies as blue, unrevivable bodies as dark blue. It has a mode selection interface."
 	icon_state = "minidetector"
 	icon = 'icons/obj/items/guns/attachments/rail.dmi'
 	slot = ATTACHMENT_SLOT_RAIL
 	attachment_action_type = /datum/action/item_action/toggle
-	/// Who's using this item
-	var/mob/living/carbon/human/operator
-	///If a hostile was detected
-	var/hostile_detected = FALSE
-	///The time needed after the last move to not be detected by this motion detector
-	var/move_sensitivity = 1 SECONDS
-	///The range of this motion detector
-	var/range = 16
-	///The list of all the blips
-	var/list/obj/effect/blip/blips_list = list()
+	active = FALSE
 
 /obj/item/attachable/motiondetector/Destroy()
-	clean_operator(forced = TRUE)
+	if(active)
+		deactivate()
 	return ..()
 
+/obj/item/attachable/motiondetector/update_icon_state()
+	. = ..()
+	icon_state = initial(icon_state) + (!active ? "" : "_on")
+
 /obj/item/attachable/motiondetector/activate(mob/user, turn_off)
-	if(operator)
-		clean_operator(forced = TRUE)
-		return TRUE
-	operator = user
-	RegisterSignals(operator, list(COMSIG_QDELETING, COMSIG_GUN_USER_UNSET), PROC_REF(clean_operator))
-	RegisterSignals(src, list(COMSIG_ITEM_EQUIPPED_TO_SLOT, COMSIG_ITEM_REMOVED_INVENTORY), PROC_REF(clean_operator))
-	UnregisterSignal(operator, COMSIG_GUN_USER_SET)
-	START_PROCESSING(SSobj, src)
+	if(active)
+		deactivate(user)
+		return
+	var/datum/component/motion_detector/motcomp = user.GetComponent(/datum/component/motion_detector)
+	motcomp = user.AddComponent(/datum/component/motion_detector)
+	RegisterSignal(user, COMSIG_COMPONENT_REMOVING, PROC_REF(deactivate_component))
+	active = TRUE
+	update_icon()
+
+/// Turns off / reverts everything that comes with activating it.
+/obj/item/attachable/motiondetector/proc/deactivate(mob/user)
+	var/datum/component/motion_detector/motcomp = user.GetComponent(/datum/component/motion_detector)
+	if(motcomp)
+		UnregisterSignal(user, COMSIG_COMPONENT_REMOVING)
+		qdel(motcomp)
+	active = FALSE
 	update_icon()
-	return TRUE
 
-///Activate the attachement when your are putting the gun out of your suit slot
-/obj/item/attachable/motiondetector/proc/start_processing_again(datum/source, obj/item/weapon/gun/equipping)
+/// As the result of the removed component, turns off / reverts everything that comes with activating it.
+/obj/item/attachable/motiondetector/proc/deactivate_component(datum/source, datum/component/removed_component)
 	SIGNAL_HANDLER
-	if(equipping != master_gun)
+
+	var/datum/component/motion_detector/motcomp = GetComponent(/datum/component/motion_detector)
+	if(!motcomp || removed_component != motcomp)
 		return
```

Comment:
wtf? why is this not handling like the autodoc component? why are you putting it this hackily on the mob?

Reply:
You're also inconsistently using getcomponent on the user and mob? have you even tested any of this?